keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1285 @@
+could put significant overhead on fetching the Job object data from etcd.
+
+In order to mitigate this risk we constrain the maximal number of failed indexes
+stored in the `.status.failedIndexes` field to `10^5` by:
+- limiting `.spec.completions` to `10^5` if `spec.runPolicy.maxFailedIndexes` is
+  not specified
+- `spec.runPolicy.maxFailedIndexes` if specified; and the max number of failed
+  indexes is limited by `10^5`.
+
+With this limit the worst-case scenario is when every second index at the end
+of completions fails, so the last 2*10^5 indexes are interleaving between
+`.status.completedIndexes` and `.status.failedIndexes`. In that case, assuming
+completions is 10^6 and maxFailedIndexes=10^5, the size of the
+`.status.completedIndexes` is limited by 700KB, and similarly the size of the
+`.status.failedIndexes` field is limited by 700KB. Together the two fields
+are limited by `1.4MB`, which is less than the etcd limit of `1.5MB`.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+We introduce a new Job API field, called `.spec.runPolicy.backoffLimitPerIndex`.
+When set to `true` it indicates that the `spec.backoffLimit` field is used to
+limit the number of retries, counted independently for all indexes.
+
+Additionally, we propose the `.spec.runPolicy.maxFailedIndexes` to control
+the maximal number of failed indexes. Once the number is exceeded the entire
+Job is marked Failed and its execution is terminated.
+
+We also propose to extend the PodFailurePolicy with a new action, called
+`FailIndex` to allow an index to fail fast before reaching the backoff limit
+per index.
+
+### Job API
+
+```golang
+
+// PodFailurePolicyAction specifies how a Pod failure is handled.
+// +enum
+type PodFailurePolicyAction string
+
+const (
+  // This is an action which might be taken on a pod failure - mark the
+  // Job's index as failed to avoid restarts within this index. This action
+  // can only be used when backoffLimitPerIndex=true.
+  PodFailurePolicyActionFailIndex PodFailurePolicyAction = "FailIndex"
+  ...
+)
+...
+
+// JobRunPolicy describes custom policy for running a Job
+type JobRunPolicy struct {
+
+  // Specifies if backoffLimit is used to limit the number of retries within an
+  // index before marking this index as failed. When enabled the number of
+  // failures per index is kept in the pod's
+  // batch.kubernetes.io/job-index-failure-count annotation. It can only
+  // be set to true when Job's completionMode=Indexed, and the Pod's restart
+  // policy is Never.
+  // The value is immutable.
+  // Defaults to false
+  // +optional
+  BackoffLimitPerIndex *bool
```

Comment:
I would consider copying the `BackoffLimit` field in this struct, and requiring in validation that only one of them is set. This can serve backwards-compatibility while making users of `runPolicy` have everything in the same place.

There might be a concern about control-plane skew or version rollbacks. We can still handle this by taking the value for backoffLimit from either place, regardless of the feature gate. This is acceptable because it's not new functionality.

Reply:
It can be done, but we need to be careful. For example, say someone sets .spec.runPolicy.backoffLimit=5 (leaving .spec.backoffLimit=nil) on 1.28 (alpha) and downgrades. Then, the Job controller crushes, because it expects the .spec.backoffLimit to be set.

Also, backoffLimit is mutable, so there might be some controllers manipulating it, that would need to now first detect which backoffLimit is set.

Personally, I don't see much of a benefit from copying the field into the new struct, but let's see what others think.

Reply:
> It can be done, but we need to be careful. For example, say someone sets .spec.runPolicy.backoffLimit=5 (leaving .spec.backoffLimit=nil) on 1.28 (alpha) and downgrades. Then, the Job controller crushes, because it expects the .spec.backoffLimit to be set.

You probably shouldn't downgrade a cluster with alpha features enabled :)
But if you do so, job controller won't crash, because we set the defaults when we read from etcd. Then what the job controller will see is the default: 6. This is acceptable, as the user was already trying to use an alpha field.

> Also, backoffLimit is mutable, so there might be some controllers manipulating it, that would need to now first detect which backoffLimit is set.

Yeah, that might be more problematic. For example, a controller using v1.27 APIs, running on a 1.29 cluster.

Reply:
Let's not copy the `BackoffLimit` here.  If we ever created a v2 job (I'm not pitching for this at this moment), it probably makes sense to partition different types of jobs: indexed jobs appear to need different concepts for error failure handling.

Comparing two possible API options, what are the pros/cons of adding a new field for `PerIndexPodFailureCount` as opposed to changing the meaning of an existing field `BackoffLimit`?  Adding a new field avoids different handling in skew scenarios and allows for composing the features in meaningful ways.

Reply:
@deads2k thanks for looking into the KEP. 

We have considered the new *int32 field (with a tentative name `backoffLimitPerIndex`) that would work along with `backoffLimit`. This is mentioned in the Alternatives `Enforce backoffLimit and backoffLimitPerIndex` section.

The **pros** of a new `int32` field:
- allows for `backoffLimit` and `PerIndexPodFailureCount` to be enforced independently, giving more flexibility

The **cons** of the `int32` field:
- can be confusing to users. Having backoffLimit defaulting to 6 may result in unexpected Job terminations when `PerIndexPodFailureCount` is set. In most use cases, the user will only want to have `PerIndexPodFailureCount`. In that case we have two options:
  1. Set backoffLimit >= `completions` * `PerIndexPodFailureCount`, which requires non-obvious calculations from the user
  2. Leave backoffLimit nil (interpreted and infinity), we need to allow for this in validation, but in that case the Job controller would break on downgrade, as the current Job controller would panic consuming a Job with backoffLimit=nil.

Also, the flexibility mentioned as a pros can be achieved by combining `backoffLimitPerIndex=true` with `maxFailedIndexes`, to control both the number of failures within an index, and the total number of failed indexes. I

Reply:
I had an offline discussion with @deads2k and we believe the cons can be mitigated by following the approach in (ii) - backoffLimit=nil interpreted as infinity. 

My concern about the Job controller panic on downgrade isn't valid as the defaulting is also triggered on reads. So the Job controller would get the defaulted Job object. 

Moreover, the approach of having two fields would allow for a more controlled downgrade, as the user could specify both `backoffLimit` and `backoffLimitPerIndex` before downgrading. 

Still, in majority of cases setting both fields can be confusing so the idea is to validate such configuration as a warning. I'm about to update the KEP accordingly.

Reply:
@deads2k @alculquicondor Updated the KEP to use the *int32 field. PTAL.

Reply:
Resolving  the conversation. As agreed on a different thread the new fields are added at the top-level for the Job spec v1. We can consider different organization for v2 .